CON-472/473: Add wrapper classes to separate blocks from ballots.

goral09 opened the pull request:
### Overview
_Provide a brief description of what this PR does, and why it's needed._

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/CON-472
https://casperlabs.atlassian.net/browse/CON-473
(and partially)
https://casperlabs.atlassian.net/browse/CON-474
https://casperlabs.atlassian.net/browse/CON-475
https://casperlabs.atlassian.net/browse/CON-476

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [ ] If this PR adds a new feature, it includes tests related to this feature.
- [ ] You assigned one person to review this PR.
- [ ] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.

### Notes
CON-{474,475,476} are marked as _partially done_ as new `Message` ADT is now being used in the fork choice, equivocation detector, and finality. In the scope of these tasks, we need to make a second pass and add tests showing that indeed ballots are properly accounted for in these 3 places.

goral09 commented:
bors try

goral09 commented:
bors retry

goral09 commented:
bors retry

goral09 commented:
bors retry

goral09 commented:
bors try

goral09 commented on casper/src/main/scala/io/casperlabs/casper/MultiParentCasperImpl.scala:
```diff
@@ -534,9 +536,6 @@ class MultiParentCasperImpl[F[_]: Sync: Log: Metrics: Time: FinalityDetector: Bl
             )
-      dag               <- dag
-      justificationMsgs <- justifications.toList.traverse(j => dag.lookup(j.latestBlockHash))
-      rank              = ProtoUtil.calculateRank(justificationMsgs.flatten)
```
Removed this traversal as `rank` is calculated in a method before that, it only needed to be passed in here.

goral09 commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -181,16 +185,6 @@ object ProtoUtil {
   def weightMapTotal(weights: Map[ByteString, Long]): Long =
     weights.values.sum
 
-  def minTotalValidatorWeight[F[_]: Monad](
```
That method was ununused.

goral09 commented on casper/src/test/scala/io/casperlabs/casper/helper/BlockGenerator.scala:
```diff
@@ -35,28 +34,19 @@ object BlockGenerator {
+          )
     } yield b
 
-  def computeBlockCheckpoint[F[_]: Sync: BlockStorage: DeployStorage: ExecutionEngineService: Log](
```
Inlined the calls to this method and removed as it was not used  anymore.

goral09 commented:
@aakoshh @AbnerZheng Can I ask you guys to do a preliminary pass over this PR? It is in the mostly final state.

AbnerZheng commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -62,8 +63,11 @@ object ProtoUtil {
-      result                <- isInMainChain(dag, candidateBlockSummary.get, targetBlockHash)
+      messageSummary <- dag.lookup(candidateBlockHash).map(_.get)
+      result <- messageSummary match {
+                 case _: MBallot => false.pure[F]
```
Could you please add a comment explaining why returning false directly?

goral09 replied:
Sure, good catch. I was thinking about this today, after reading the spec and I hesitate now whether to not take ballot's target hash and check whether it is in the main chain instead.

goral09 replied:
@AbnerZheng look at the last commit.

goral09 replied:
Actually, I've changed my mind again and I think that the original version was a correct one. To _be in a main chain_ means to be a main child but ballots are not children. They are not even part of the p-DAG.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/equivocations/EquivocationDetector.scala:
```diff
@@ -90,9 +90,9 @@ object EquivocationDetector {
+                            implicit0(blockTopoOrdering: Ordering[MessageSummary]) = DagOperations.blockTopoOrderingDesc
                             stream = DagOperations.bfToposortTraverseF(
-                              List(BlockSummary.fromBlock(block))
+                              MessageSummary.fromBlock(block).toList
```
`fromBlock(...).toList` looks strange.

goral09 replied:
Especially this is not a test. I will rewrite that. Thanks for spotting.

goral09 replied:
Done.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/finality/votingmatrix/VotingMatrix.scala:
```diff
@@ -40,7 +42,18 @@ object VotingMatrix {
+      block <- message.get match {
+                case b: MBlock => b.pure[F]
+                case ballot: MBallot =>
+                  MonadThrowable[F].raiseError[MBlock](
```
I think these `MBlock` and `MBallot` types look worse than if they were `BlockMessage` and `BallotMessage`.

goral09 replied:
`BlockMessage` is already taken right?

aakoshh replied:
Or `Message.Block` and `Message.Ballot`.

goral09 replied:
The number of places I will have to make this change to avoid conflicts :man_facepalming:

aakoshh replied:
Ah, in the `protocol` namespace. Well maybe `Message.Block` then, but leave it you like it. It feels strange to try to vocalize :D

goral09 replied:
I did refactor it to be `Message.Block` and `Message.Ballot`. Will push in a moment.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -30,10 +31,10 @@ object ProtoUtil {
       targetBlockHash: BlockHash
   ): F[Boolean] =
-    if (candidateBlockSummary.blockHash == targetBlockHash) {
+    if (candidateBlockSummary.messageHash == targetBlockHash) {
```
If we know that it's the block version, not the ballot, I'd use `blockHash`.

goral09 replied:
that field is from the `Message` class, not block.

aakoshh replied:
`candidateBlockSummary: MBlock,`, so I thougth we know it's a block. But it doesn't matter that much.

goral09 replied:
We know it's a block but the field `messageHash` is inherited from the `Message`.

aakoshh replied:
It could have a `blockHash` variant. But again, never mind.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -62,8 +63,11 @@ object ProtoUtil {
-      result                <- isInMainChain(dag, candidateBlockSummary.get, targetBlockHash)
+      messageSummary <- dag.lookup(candidateBlockHash).map(_.get)
+      result <- messageSummary match {
+                 case b: MBallot => isInMainChain(dag, b.parentBlock, targetBlockHash)
```
Can a parent of a ballot not be another ballot, if the validator has no block to produce for a long time?

goral09 replied:
No. Ballot always targets another block.

aakoshh replied:
Correct me if I'm wrong, but then when we download a DAG, you are not even supposed to see the ballots?

goral09 replied:
Ballots have to be gossiped, just like blocks.

aakoshh replied:
Okay I see they have to form a valid j-DAG, only the parent points back to the last actual block. This is what makes me think we should filter them out in `showBlocks`. Except that one works on ranks and would show empty pages if there are no blocks. Okay never mind.

goral09 replied:
I don't have an opinion about `showBlock`. Ballots are not very interesting to the dApp developers. They are kind of internal machinery to finalize blocks.

aakoshh replied:
We may have to rework showBlocks to dynamically find X ranks of blocks which are actual blocks, not just ballots.

goral09 commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -62,8 +63,11 @@ object ProtoUtil {
     for {
-      candidateBlockSummary <- dag.lookup(candidateBlockHash)
-      result                <- isInMainChain(dag, candidateBlockSummary.get, targetBlockHash)
+      messageSummary <- dag.lookup(candidateBlockHash).map(_.get)
```
First, if you look closely I haven't changed this function, that `.get` was there earlier. Second,  this method is never called by the block API.

aakoshh replied:
It's not called now, but there used to be some main chain variant of show blocks. Anyway, fine with me.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -249,14 +233,36 @@ object ProtoUtil {
+                s"A ballot ${PrettyPrinter.buildString(b.messageHash)} was a parent block for ${PrettyPrinter
+                  .buildString(message.messageHash)}"
+              )
+            )
```
I actually thought Ballots are like empty blocks that show up in the DAG. Otherwise how will you download them later, if they are not part of justifications?

aakoshh replied:
Or is it that they can be justifications, but not parents?

goral09 replied:
> Or is it that they can be justifications, but not parents?

That :point_up:

aakoshh replied:
Do my ballot justifications form a chain then?

goral09 replied:
Well, they do have to include one another in the justifications (otherwise you equivocate).

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/validation/ValidationImpl.scala:
```diff
@@ -403,37 +401,45 @@ class ValidationImpl[F[_]: MonadThrowable: FunctorRaise[?[_], InvalidBlock]: Log
+                                            case (_, Justification(_, latestBlockHash)) =>
+                                              dag.lookup(latestBlockHash).flatMap {
+                                                case Some(meta) =>
+                                                  meta.validatorMsgSeqNum.pure[F]
```
Do ballots increase the block sequence number as well?

aakoshh replied:
The definition of equivocation is now having more than 1 block|ballot with the same seq. nr?

goral09 replied:
The definition of equivocation is that two messages from the same validator don't include each other in their justification. The fact that we have a validator sequence number is an optimization that we leverage. So, to answer your first question - yes.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+) extends MessageSummary {
+  // For Genesis block we expect it to have no parents.
+  // We could either encode it as separate ADT variant or keep the assumptions.
+  override val parents: Seq[Id] = (parentBlock +: secondaryParents).filterNot(_ == ByteString.EMPTY)
```
Yeah this doesn't feel right, to serve empty as `parentBlock`. I can imagine you pass it here and there and eventually get an error. Better would be to make it an options, or as you say make separate case for the Genesis block, which doesn't have justifications either.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+      val justifications     = header.justifications
+      val rank               = header.rank
+      val validatorMsgSeqNum = header.validatorBlockSeqNum
+      val messageType        = header.roleType
```
Pff, I whoudl have noticed this, but surely I thought you'd just name the field `role`.

goral09 replied:
Message type just sounds better :P I will rename :)

aakoshh replied:
I'm talking about the `role`, but you know what, at this point the whole codebase is full of this MessageSummary, so if you want just name it `message_type` in proto. I didn't think this is what you'll name your model.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+            MBallot(
+              messageHash,
+              creator,
+              parentBlock,
```
ValidationImpl should check that the `BlockSummary` really only had 1 parent.

goral09 replied:
I agree it should do that and a lot more w.r.t to these ballots vs blocks. Ideally we would integrate it somehow but I thought this is too much work for one PR. I will add those validations though.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+          )
+        case BLOCK =>
+          val secondaryParents =
+            if (header.parentHashes.isEmpty) Seq.empty else header.parentHashes.tail
```
The `tail` of an empty list is empty as well, isn't it? Do you need to check?

goral09 replied:
`tail` on an empty list throws an error.

aakoshh replied:
Hah! Forgot that. Thanks

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+        case Unrecognized(_) =>
+          Failure(
+            new IllegalArgumentException(
+              s"A message ${Base16.encode(messageHash.toByteArray).take(10)} has unrecognized message type."
```
```suggestion
              s"Block ${Base16.encode(messageHash.toByteArray).take(10)} has unrecognized role: $other."
```

goral09 replied:
That's the thing - it's not a block, yet. It's a message. If it was a block then its role would be a `BLOCK`.

aakoshh replied:
Well it was a `Block` or `BlockSummary` in the API.
Anyway, at least say "The message..." and name the value given.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+      case ex: Throwable =>
+        val message =
+          s"A message ${Base16.encode(b.blockHash.toByteArray).take(10)} could not be parsed to either Block type or a Ballot."
+        Failure(new IllegalArgumentException(message, ex))
```
In the normal workflow (i.e. when not deserializing from storage), is this happening before or after validation?

goral09 replied:
After. Actually, now it only ever happens when we read data from the `DagRepresentation`.

aakoshh commented on models/src/main/scala/io/casperlabs/models/MessageSummary.scala:
```diff
@@ -0,0 +1,125 @@
+            if (header.parentHashes.isEmpty) Seq.empty else header.parentHashes.tail
+          val weightMap = header.getState.bonds.map {
+            case Bond(validatorPk, stake) => validatorPk -> stake
+          }.toMap
```
I wonder if you wouldn't be better off making these a lazy value in the subclasses. I imagine a lot of times you won't need them at all. Or you want to really want to exercise all `getXXX` method here at the cost of performance?

goral09 replied:
Making it a `lazy val` is a good idea.

AbnerZheng commented on storage/src/test/scala/io/casperlabs/storage/dag/IndexedDagStorage.scala:
```diff
@@ -46,7 +45,7 @@ final class IndexedDagStorage[F[_]: Monad](
       nextCreatorSeqNum <- dag
                             .latestMessage(block.getHeader.validatorPublicKey)
-                            .map(_.fold(-1)(_.validatorBlockSeqNum) + 1)
+                            .map(_.fold(0)(_.validatorMsgSeqNum) + 1)
```
This will make `validatorBlockSeqNum`  of genesis block to be 1 in test, which is not the same as PROD.

goral09 replied:
Not really. Genesis will get `0` but we are here interested in the _next validator sequence number_. So, for the very first block from this validator it will be 1.

aakoshh commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -126,13 +128,17 @@ class SQLiteDagStorage[F[_]: Bracket[?[_], Throwable]](
+      .flatMap {
+        case None     => (None: Option[MessageSummary]).pure[F]
+        case Some(bs) => toMessageSummaryF(bs).map(Some(_))
+      }
```
Looks like this boilderplate could be move to `MessageSummary`

goral09 replied:
What? Raising to `F` context? Or handling _optional_ block summary?

aakoshh replied:
Both.

goral09 replied:
I thought about it but it felt really strange and artificial :/ I didn't like it so I didn't do it. If you insist I can add that though.

goral09 replied:
Done.

aakoshh replied:
Thanks. I just see it repeated here quite a few times and it feels like the storage class should spend as little code on this mapping as possible, so that's it's as easy and painless to migrate as you can make it.

aakoshh commented on storage/src/test/scala/io/casperlabs/storage/dag/IndexedDagStorage.scala:
```diff
@@ -46,7 +45,7 @@ final class IndexedDagStorage[F[_]: Monad](
       nextCreatorSeqNum <- dag
                             .latestMessage(block.getHeader.validatorPublicKey)
-                            .map(_.fold(-1)(_.validatorBlockSeqNum) + 1)
+                            .map(_.fold(0)(_.validatorMsgSeqNum) + 1)
```
Yay :+1:

AbnerZheng commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -189,14 +195,18 @@ class SQLiteDagStorage[F[_]: Bracket[?[_], Throwable]](
       .option
       .transact(xa)
+      .flatMap {
+        case None     => (None: Option[MessageSummary]).pure[F]
```
how about `none[MessageSummary].pure[F]`?

goral09 replied:
Great idea.

aakoshh commented:
Do you think `showBlocks` should have an option to include/exclude ballots?
What should be the default?

goral09 commented:
@aakoshh @AbnerZheng I've addressed all of your comments except one (https://github.com/CasperLabs/CasperLabs/pull/1103#discussion_r325087539). This will have to wait until tomorrow morning as I have more pressing issues at hand right now.

aakoshh commented on models/src/main/scala/io/casperlabs/models/Message.scala:
```diff
@@ -0,0 +1,133 @@
+      val justifications     = header.justifications
+      val rank               = header.rank
+      val validatorMsgSeqNum = header.validatorBlockSeqNum
+      val role               = header.roleType
```
In the next PR, could you rename this field to `header.role` or `header.messageType`?

goral09 replied:
Will do in current PR.

aakoshh commented on models/src/main/scala/io/casperlabs/models/Message.scala:
```diff
@@ -0,0 +1,133 @@
+      val messageHash        = b.blockHash
+      val header             = b.getHeader
+      val parentBlock        = header.parentHashes.headOption.getOrElse(ByteString.EMPTY)
+      val creator            = header.validatorPublicKey
```
I think I like `validatorId` better but I see you must have been looking at Abner's PR for a long time, you have been assimilated.

goral09 replied:
Will rename :)

aakoshh commented on models/src/main/scala/io/casperlabs/models/Message.scala:
```diff
@@ -0,0 +1,133 @@
+
+object Message {
+  case class Block private (
+      messageHash: Message#Id,
```
I guess this could be `Message.Id` if you just moved the `type Id` alias into the companion.

aakoshh replied:
Okay. Not sure why you'd want to hide something like the ID which is used across all interfaces to look up these records and to provide references to them though.

aakoshh commented on models/src/main/scala/io/casperlabs/models/Message.scala:
```diff
@@ -0,0 +1,133 @@
+  * This way we can do necessary validation of protobuf fields (like existence of required fields)
+  * and represent different messages in type-safe way.
+  */
+abstract class Message {
```
Since there's no constructor or anything, might as well make it into a `trait Message`

goral09 replied:
I always thought that traits are for defining behavior/API and classes encapsulate state.

aakoshh replied:
Since you modelled the Block and Ballot with case classes they can't take advantage of inheritance via base class constructors, which is one of the things an abstract class can have that a trait cannot. In this case the Message provides the API, but no state.

goral09 replied:
OK.

aakoshh commented on models/src/main/scala/io/casperlabs/models/BlockImplicits.scala:
```diff
@@ -51,4 +51,11 @@ object BlockImplicits {
+      msg.parents.isEmpty &&
+        msg.validatorId.isEmpty &&
+        msg.signature.sig.isEmpty
+  }
```
Shouldn't this be just moved into the `Message` type itself?

goral09 replied:
Oh right, good observation.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/validation/Validation.scala:
```diff
@@ -42,6 +42,9 @@ trait Validation[F[_]] {
   def formatOfFields(b: BlockSummary, treatAsGenesis: Boolean = false): F[Boolean]
 
+  // Validates that a message that is supposed to be a ballot adheres to ballot's specification.
+  def ballot(b: BlockSummary): F[Unit]
```
Not every field is exposed on the trait now, to make it easier to navigate what needs to be called and what doesn't. Since you are calling this inside `Validation.blockSummary` I think it's fine to omit it from `Validation`.

goral09 replied:
To omit that a message that claims to be a ballot actually adheres to the specification of a ballot?

goral09 replied:
Ah, I see. You want me to remove it from the trait and leave in the `ValidationImpl`. Will do that. I think that `Validation` being a trait is wrong anway.

aakoshh replied:
Yeah, it was supposed to help with simulation somehow.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/util/ProtoUtil.scala:
```diff
@@ -249,14 +235,36 @@ object ProtoUtil {
+        /** We know that Gensis is of [[Message.Block]] type */
+        message.asInstanceOf[Message.Block].weightMap.pure[F]
+      } else {
+        dag.lookup(message.parentBlock).flatMap {
```
If accessing the main parent is not very often (unlike `parents` which can obviously be empty), I'd consider differentiating the `Message.Genesis` and `Message.ValidatorMessage` traits, where only the latter would have a parent field.

goral09 replied:
We access main parent more often than parents because the latter almost don't matter. This is also exactly what we talked about earlier today. If I will do that then dozens of tests will start to fail and I don't  have days to fix them

goral09 replied:
Also, this won't change anything (at least in this particular place) because Genesis could still be returned here and we have to handle it

aakoshh replied:
Okay, I didn't know it was so wide spread. So basically the tests used something like `.parentHashes.head` all the time because they knew they weren't dealing with Genesis.

goral09 commented:
bors try

goral09 commented:
bors try

goral09 commented:
bors retry

goral09 commented:
bors retry

goral09 commented:
bors retry

goral09 commented:
bors retry

goral09 commented:
bors try

goral09 commented:
bors retry

goral09 commented:
bors try
